src/core/context/AppContextProvider.tsx

```diff
@@ -146,10 +146,45 @@ export const AppContextProvider = (props: {
     [conversations],
   );
 
-  const handleModelChange = useCallback((modelName: SupportedModels) => {
-    const newConfig = getModelConfig(modelName);
-    setModelConfig(newConfig);
-  }, []);
+  const ensureCorrectSystemPrompt = useCallback(
+    (messageStore: MessageHistoryStore, modelConfig: ModelConfig) => {
+      const messages = messageStore.getSnapshot();
+      const systemPrompt = modelConfig.systemPrompt;
+
+      // If no messages, add the system message
+      if (messages.length === 0) {
+        messageStore.addMessage({
+          role: 'system',
+          content: systemPrompt,
+        });
+      } else if (messages[0].role === 'system') {
+        // If first message is system, update it
+        const updatedMessages = [...messages];
+        updatedMessages[0] = {
+          role: 'system',
+          content: systemPrompt,
+        };
+        messageStore.setMessages(updatedMessages);
+      }
+    },
```

Comment:
Would it be overkill to also check for `messages.length === 1` here? Shouldn't we only ever be resetting the system prompt (and then the messageStore) if its empty or only has a system prompt? Spreading existing messages after an updated system prompt wouldn't actually happen, right?

```
} else if (messages[0].role === 'system' && messages.length === 1) {
  const updatedMessages = [ {
    role: 'system',
    content: systemPrompt,
  }];
  messageStore.setMessages(updatedMessages);
}
      ```